odl/space/npy_tensors.py

```diff
@@ -0,0 +1,1992 @@
+        False
+
+        Space membership matters:
+
+        >>> space2 = odl.tensor_space(3, dtype='int64')
+        >>> y = space2.element([1, 2, 3])
+        >>> x == y or y == x
+        False
+        """
+        if other is self:
+            return True
+        elif other not in self.space:
+            return False
+        else:
+            return np.array_equal(self.data, other.data)
+
+    def copy(self):
+        """Create an identical (deep) copy of this vector.
+
+        Parameters
+        ----------
+        None
+
+        Returns
+        -------
+        copy : `NumpyTensor`
+            The deep copy
+
+        Examples
+        --------
+        >>> space = odl.rn(3)
+        >>> x = space.element([1, 2, 3])
+        >>> y = x.copy()
+        >>> y == x
+        True
+        >>> y is x
+        False
+        """
+        return self.space.element(self.data.copy())
+
+    __copy__ = copy
+
+    def __getitem__(self, indices):
+        """Return ``self[indices]``.
+
+        Parameters
+        ----------
+        indices : index expression
+            Integer, slice or sequence of these, defining the positions
+            of the data array which should be accessed.
+
+        Returns
+        -------
+        values : `NumpyTensorSpace.dtype` or `NumpyTensor`
+            The value(s) at the given indices. Note that the returned
+            object is a writable view into the original tensor.
+
+        Examples
+        --------
+        >>> space = odl.rn(3)
+        >>> x = space.element([1, 2, 3])
+        >>> x[0]
+        1.0
+        >>> x[1:]
+        rn(2).element([2.0, 3.0])
+        >>> space = odl.rn((2, 3))
+        >>> x = space.element([[1, 2, 3],
+        ...                    [4, 5, 6]])
+        >>> x[0, 1]
+        2.0
+        >>> x[:, 1:]
+        rn((2, 2)).element(
+            [[2.0, 3.0],
+             [5.0, 6.0]]
+        )
+        >>> y = x[[[0, 1], [1, 2]]]
+        >>> y
+        rn(2).element([2.0, 6.0])
+        >>> y[:] = 0
+        >>> x  # not modified
+        rn((2, 3)).element(
+            [[1.0, 2.0, 3.0],
+             [4.0, 5.0, 6.0]]
+        )
+
+        Slices can be assigned to, except if lists are used for
+        indexing:
+
+        >>> y = x[:, ::2]  # view into x
+        >>> y[:] = -9
+        >>> x
+        rn((2, 3)).element(
+            [[-9.0, 2.0, -9.0],
+             [-9.0, 5.0, -9.0]]
+        )
+
+        """
+        arr = self.data[indices]
+        if np.isscalar(arr):
+            return arr
+        else:
+            space_constructor = type(self.space)
+            if (self.order in ('C', 'F') and
+                    arr.flags[self.order + '_CONTIGUOUS']):
+                out_spc_order = self.order
+            else:
+                # To preserve the array view for non-contiguous slices,
+                # we need to use 'A' for the space in that case.
+                out_spc_order = 'A'
+            space = space_constructor(
```

Comment:
Not really, at least not as long as `space.__getitem__` really captures all the relevant cases. There's much more going on with Numpy indexing that would be very tedious to re-implement. I like this lazy approach much more, since `arr[indices]` tells us everything we need to know.

As I wrote above, I really see rather little use for `space.__getitem__` unless it really does what you write here.